Using Effects in query-json (draft) (#747)
davesnx/sancho.dev: src/content/posts/effect-in-query-json.mdx

If you know how `try/catch` works in any other language, you already have the core intuition for this post. Effects generalize that model in exactly one way, and that one difference is the whole story.

If you have never used jq: a jq filter is an expression that takes JSON in and produces JSON out. `.name` extracts a field, `.users[]` iterates over an array, and `|` pipes the output of one filter into the next as its input. query-json implements that same filter language.

Since then, query-json grew into most of jq's language: `try/catch`, `while`, `limit`, user-defined functions, and expressions that produce zero, one, or many results. Those expressions can fail, halt the program, break out of loops, and carry the rest of the control-flow repertoire you would expect from a complete programming language. jq is, in fact, [Turing-complete](https://github.com/MakeNowJust/bf.jq) (someone wrote a Brainfuck interpreter in it, if you need convincing). The design I started with could not stretch to cover any of it.

> This post is about how the interpreter evolved to handle those control-flow patterns, and the part OCaml's effect system played in it.

Reviewer: something like this?

```suggestion
and how I used OCaml's effect system to implement it
```

## What makes jq a real language

> Most jq expressions are **generators**: a single input can produce multiple outputs. That one property, described in the [jq manual](https://jqlang.org/manual/), shapes the entire interpreter.

Reviewer: I would remove "That one property, described in the [jq manual](https://jqlang.org/manual/), shapes the entire interpreter."
